File: backend/Zeladoria.Application/UseCases/Auth/LoginUseCase.cs
using System;
using System.Threading.Tasks;
using Zeladoria.Application.DTOs;
using Zeladoria.Application.Interfaces;
using Zeladoria.Application.Utils;
using Zeladoria.Domain.Interfaces;

namespace Zeladoria.Application.UseCases.Auth;

/// <summary>
/// Caso de uso responsável por realizar o login de usuários no sistema.
/// Encapsula a lógica de negócio para validação de credenciais e geração de JWT.
/// </summary>
public class LoginUseCase
{
    private readonly IUserRepository _userRepository;
    private readonly IJwtService _jwtService;

    public LoginUseCase(IUserRepository userRepository, IJwtService jwtService)
    {
        _userRepository = userRepository;
        _jwtService = jwtService;
    }

    /// <summary>
    /// Executa o fluxo principal de autenticação.
    /// </summary>
    /// <param name="input">Dados de entrada do usuário (CPF e Senha).</param>
    /// <returns>DTO contendo o Token JWT gerado e os dados de perfil do usuário.</returns>
    /// <exception cref="Exception">Lançada caso o CPF não seja encontrado ou a senha esteja incorreta.</exception>
    public async Task<AuthOutputDto> ExecuteAsync(LoginInputDto input)
    {
        var user = await _userRepository.GetByCpfAsync(input.Cpf);

        if (user == null || !AuthUtils.VerifyPassword(input.Password, user.PasswordHash))
        {
            throw new Exception("CPF ou senha inválidos.");
        }

        var token = _jwtService.GenerateToken(user);

        return new AuthOutputDto
        {
            Token = token,
            Name = user.Name,
            Email = user.Email,
            Cpf = user.Cpf,
            Role = user.Role,
            UserId = user.Id.ToString(),
            CreatedAt = user.CreatedAt
        };
    }
}

File: backend-java/src/main/java/br/com/fiap/hackgov/application/usecase/auth/LoginUseCase.java
package br.com.fiap.hackgov.application.usecase.auth;

import br.com.fiap.hackgov.application.dto.auth.AuthOutputDto;
import br.com.fiap.hackgov.application.dto.auth.LoginInputDto;
import br.com.fiap.hackgov.application.service.JwtService;
import br.com.fiap.hackgov.application.util.AuthUtils;
import br.com.fiap.hackgov.domain.entity.User;
import br.com.fiap.hackgov.domain.repository.UserRepository;
import org.springframework.stereotype.Service;

@Service
public class LoginUseCase {

    private final UserRepository userRepository;
    private final JwtService jwtService;

    public LoginUseCase(UserRepository userRepository, JwtService jwtService) {
        this.userRepository = userRepository;
        this.jwtService = jwtService;
    }

    public AuthOutputDto execute(LoginInputDto input) {
        User user = userRepository.getByCpf(input.cpf())
                .orElseThrow(() -> new IllegalArgumentException("CPF ou senha inválidos."));

        if (!AuthUtils.verifyPassword(input.password(), user.getPasswordHash())) {
            throw new IllegalArgumentException("CPF ou senha inválidos.");
        }

        String token = jwtService.generateToken(user);

        return new AuthOutputDto(
                token,
                user.getName(),
                user.getEmail(),
                user.getCpf(),
                user.getRole(),
                user.getId(),
                user.getCreatedAt()
        );
    }
}
